shaping/input.go

```diff
@@ -33,3 +35,85 @@ type Input struct {
 	// Language is an identifier for the language of the text.
 	Language language.Language
 }
+
+// SplitByFontGlyphs split the runes from 'input' to several items, sharing the same
+// characteristics as 'input', expected for the `Face` which is set to
+// the first font among 'availableFonts' providing support for all the runes
+// in the item.
+// Runes supported by no fonts are mapped to the first element of 'availableFonts', which
+// must not be empty.
+// The 'Face' field of 'input' is ignored: only 'availableFaces' are consulted.
+// Rune coverage is obtained by calling the NominalGlyph() method of each font.
+func SplitByFontGlyphs(input Input, availableFaces []font.Face) []Input {
+	var splittedInputs []Input
+	currentInput := input
+	for i := input.RunStart; i < input.RunEnd; i++ {
+		r := input.Text[i]
+		if currentInput.Face != nil && ignoreFaceChange(r) {
+			// add the rune to the current input
+			continue
+		}
+
+		// select the first font supporting r
+		selectedFace := availableFaces[0]
+		for _, f := range availableFaces {
+			if _, has := f.NominalGlyph(r); has {
+				selectedFace = f
+				break
+			}
+		}
+
+		if currentInput.Face == selectedFace {
+			// add the rune to the current input
+			continue
+		}
+
+		// new face needed
+
+		if i != input.RunStart {
+			// close the current input ...
+			currentInput.RunEnd = i
+			// ... add it to the output ...
+			splittedInputs = append(splittedInputs, currentInput)
+		}
+
+		// ... and create a new one
+		currentInput = input
+		currentInput.RunStart = i
+		currentInput.Face = selectedFace
+	}
+
+	// close and add the last input
+	currentInput.RunEnd = input.RunEnd
+	splittedInputs = append(splittedInputs, currentInput)
+	return splittedInputs
+}
+
+// ignoreFaceChange returns `true` is the given rune should not trigger
+// a change of font.
+//
+// We don't want space characters to affect font selection; in general,
+// it's always wrong to select a font just to render a space.
+// We assume that all fonts have the ASCII space, and for other space
+// characters if they don't, HarfBuzz will compatibility-decompose them
+// to ASCII space...
+//
+// We don't want to change fonts just for variation selectors.
+//
+// Finally, don't change fonts for line or paragraph separators.
+//
+// The rationale is taken from pango : see bugs
+// https://bugzilla.gnome.org/show_bug.cgi?id=355987
+// https://bugzilla.gnome.org/show_bug.cgi?id=701652
+// https://bugzilla.gnome.org/show_bug.cgi?id=781123
+// for more details.
```

Comment:
I'm _sooooo_ glad that you're the one who wrote this @benoitkugler! Having your experise about pango and various problems within it is proving quite invaluable! Thank you for including links to the various edge case bugs in pango.